src/main/java/com/cloudant/client/api/ClientBuilder.java

```diff
@@ -0,0 +1,531 @@
+        return this;
+    }
+
+    /**
+     * Set a custom GsonBuilder to use when serializing and de-serializing requests and
+     * responses between the CloudantClient and the server.
+     * <P>
+     * Note: the supplied GsonBuilder will be augmented with some internal TypeAdapters.
+     * </P>
+     *
+     * @param gsonBuilder the custom GsonBuilder to use
+     * @return this ClientBuilder object for setting additional options
+     */
+    public ClientBuilder gsonBuilder(GsonBuilder gsonBuilder) {
+        this.gsonBuilder = gsonBuilder;
+        return this;
+    }
+
+    /**
+     * Set the maximum number of connections to maintain in the connection pool.
+     * <P>
+     * Note: this setting only applies if using the optional OkHttp dependency. If OkHttp is not
+     * present then the JVM configuration is used for pooling. Consult the JVM documentation for
+     * the {@code http.maxConnections} property for further details.
+     * </P>
+     * Defaults to {@link #DEFAULT_MAX_CONNECTIONS}
+     *
+     * @param maxConnections the maximum number of simultaneous connections to open to the server
+     * @return this ClientBuilder object for setting additional options
+     */
+    public ClientBuilder maxConnections(int maxConnections) {
+        this.maxConnections = maxConnections;
+        return this;
+    }
+
+    /**
+     * Sets a proxy url for the client connection.
+     *
+     * @param proxyURL the URL of the proxy server
+     * @return this ClientBuilder object for setting additional options
+     */
+    public ClientBuilder proxyURL(URL proxyURL) {
+        this.proxyURL = proxyURL;
+        return this;
+    }
+
+    /**
+     * Sets an optional proxy username for the client connection.
+     *
+     * @param proxyUser username for the proxy server
+     * @return this ClientBuilder object for setting additional options
+     */
+    public ClientBuilder proxyUser(String proxyUser) {
+        this.proxyUser = proxyUser;
+        return this;
+    }
+
+    /**
+     * Sets an optional proxy password for the proxy user specified by
+     * {@link #proxyUser(String)}.
+     *
+     * @param proxyPassword password for the proxy server user
+     * @return this ClientBuilder object for setting additional options
+     */
+    public ClientBuilder proxyPassword(String proxyPassword) {
+        this.proxyPassword = proxyPassword;
+        return this;
+    }
+
+    /**
+     * Flag to disable hostname verification and certificate chain validation. This is not
+     * recommended, but for example could be useful for testing with a self-signed certificate.
+     * <P>
+     * The SSL authentication is enabled by default meaning that hostname verification
+     * and certificate chain validation is done using the JVM default settings.
+     * </P>
+     *
+     * @return this ClientBuilder object for setting additional options
+     * @throws IllegalStateException when SSLSocketFactory is not null
```

Comment:
I think this should be more about what options have been called and not what the internal check is. Maybe something like:
`@throws IllegalStateException if {@link #authenticatedModeSSLSocketFactory(SSLSocketFactory)} has been called on this ClientBuilder`